### ISSUE TITLE
Struggling with relationships

### ISSUE THREAD

**Issue by rj93-finseta** [toedter/spring-hateoas-jsonapi#110]
Hi,

I'm trying to follow along with your examples but I can't seem to get the relationships to work and not sure where I'm going wrong.

Models:

```
@NoArgsConstructor
@AllArgsConstructor
@Data
public class Movie {
    @JsonApiId
    private String id;
    private String title;
}

@NoArgsConstructor
@AllArgsConstructor
@Data
public class Director {
    @JsonApiId
    private String id;
    private String name;
}
```

I have the following controller:

```
    @RequestMapping(
        method = RequestMethod.GET,
        value = "/movies",
        produces = { "application/vnd.api+json" }
    )
    public RepresentationModel<?> movies() {
        Movie movie = new Movie("1", "Star Wars");
        Director director = new Director("1", "George Lucas");

        final RepresentationModel<?> jsonApiModel = jsonApiModel()
                .model(movie)
                .relationship("directors", jsonApiModel().model(director).build())
                .build();

        return jsonApiModel;
    }
```

However, when I attempt to run this I get:

```
jakarta.servlet.ServletException: Request processing failed: java.lang.IllegalStateException: Cannot build representation model: JSON:API relationship validation error for: com.toedter.spring.hateoas.jsonapi.JsonApiRelationship@146eda1b
...
Caused by: java.lang.IllegalStateException: Cannot build representation model: JSON:API relationship validation error for: com.toedter.spring.hateoas.jsonapi.JsonApiRelationship@146eda1b
	at com.toedter.spring.hateoas.jsonapi.JsonApiModelBuilder.build(JsonApiModelBuilder.java:608)
	at com.demo.AccountFeesController.movies(AccountFeesController.java:123)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        ...
```

If I remove `relationship("directors", jsonApiModel().model(director).build())` it works as expected. 

Where am I going wrong?

**Comment by toedter** [comment]
Just try:

``` Java
final RepresentationModel<?> jsonApiModel = jsonApiModel()
                .model(movie)
                .relationship("directors", director)
                .build();
```

**Comment by rj93-finseta** [comment]
Thanks for the quick reply, I tried that earlier but it leads to:

```
jakarta.servlet.ServletException: Request processing failed: org.springframework.http.converter.HttpMessageConversionException: Type definition error: [simple type, class com.toedter.spring.hateoas.jsonapi.JsonApiResourceIdentifier]
    ... 
Caused by: org.springframework.http.converter.HttpMessageConversionException: Type definition error: [simple type, class com.toedter.spring.hateoas.jsonapi.JsonApiResourceIdentifier]
	at org.springframework.http.converter.json.AbstractJackson2HttpMessageConverter.writeInternal(AbstractJackson2HttpMessageConverter.java:492)
	at org.springframework.http.converter.AbstractGenericHttpMessageConverter.write(AbstractGenericHttpMessageConverter.java:126)
	at org.springframework.web.servlet.mvc.method.annotation.AbstractMessageConverterMethodProcessor.writeWithMessageConverters(AbstractMessageConverterMethodProcessor.java:345)
	at org.springframework.web.servlet.mvc.method.annotation.RequestResponseBodyMethodProcessor.handleReturnValue(RequestResponseBodyMethodProcessor.java:208)
	at org.springframework.web.method.support.HandlerMethodReturnValueHandlerComposite.handleReturnValue(HandlerMethodReturnValueHandlerComposite.java:78)
	at org.springframework.hateoas.server.mvc.RepresentationModelProcessorHandlerMethodReturnValueHandler.handleReturnValue(RepresentationModelProcessorHandlerMethodReturnValueHandler.java:134)
	at org.springframework.web.method.support.HandlerMethodReturnValueHandlerComposite.handleReturnValue(HandlerMethodReturnValueHandlerComposite.java:78)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:136)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:986)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:891)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1088)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:978)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014)
	... 107 more
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Conflicting property-based creators: already had explicit creator [constructor for `com.toedter.spring.hateoas.jsonapi.JsonApiResourceIdentifier` (3 args), annotations: {interface com.fasterxml.jackson.annotation.JsonCreator=@com.fasterxml.jackson.annotation.JsonCreator(mode=DEFAULT)}, encountered another: [constructor for `com.toedter.spring.hateoas.jsonapi.JsonApiResourceIdentifier` (2 args), annotations: {interface com.fasterxml.jackson.annotation.JsonCreator=@com.fasterxml.jackson.annotation.JsonCreator(mode=DEFAULT)} (through reference chain: com.toedter.spring.hateoas.jsonapi.JsonApiDocument["data"]->com.toedter.spring.hateoas.jsonapi.JsonApiData["relationships"]->java.util.HashMap["directors"]->com.toedter.spring.hateoas.jsonapi.JsonApiRelationshipSerializer$JsonApiRelationshipForSerialization["data"])
	at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:77)
	at com.fasterxml.jackson.databind.SerializerProvider.reportBadDefinition(SerializerProvider.java:1328)
	at com.fasterxml.jackson.databind.SerializerProvider._createAndCacheUntypedSerializer(SerializerProvider.java:1475)
```

**Comment by rj93-finseta** [comment]
Maybe I should note that I'm using spring boot 3.4.0, maybe I should try with 3.3.x?

**Comment by rj93-finseta** [comment]
Just tried with a fresh start.spring.io project using Spring Boot 3.3.6 and its working, however upgrading to 3.4.0 and it then breaks. 

I realise that 3.4.0 has only just been released, so not expecting an immediate fix and currently this is just a PoC for me. Do you have an estimate for when 3.4.0 support will be added?

**Comment by toedter** [comment]
I am planning to release a version for Spring Boot 3.4. soon, maybe this weekend.

**Comment by toedter** [comment]
I just released version 2.1.1 which is based on Spring Boot 3.4.0
